hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsOutputStream.java

```diff
@@ -168,25 +199,153 @@ public AbfsOutputStream(AbfsOutputStreamContext abfsOutputStreamContext)
     this.lease = abfsOutputStreamContext.getLease();
     this.leaseId = abfsOutputStreamContext.getLeaseId();
     this.executorService =
-        MoreExecutors.listeningDecorator(abfsOutputStreamContext.getExecutorService());
+        MoreExecutors.listeningDecorator(
+            abfsOutputStreamContext.getExecutorService());
     this.cachedSasToken = new CachedSASToken(
         abfsOutputStreamContext.getSasTokenRenewPeriodForStreamsInSeconds());
     this.outputStreamId = createOutputStreamId();
-    this.tracingContext = new TracingContext(abfsOutputStreamContext.getTracingContext());
+    this.tracingContext = new TracingContext(
+        abfsOutputStreamContext.getTracingContext());
     this.tracingContext.setStreamID(outputStreamId);
     this.tracingContext.setOperation(FSOperationType.WRITE);
     this.ioStatistics = outputStreamStatistics.getIOStatistics();
     this.blockFactory = abfsOutputStreamContext.getBlockFactory();
-    this.blockSize = bufferSize;
+    this.isDFSToBlobFallbackEnabled
+        = abfsOutputStreamContext.isDFSToBlobFallbackEnabled();
     // create that first block. This guarantees that an open + close sequence
     // writes a 0-byte entry.
-    createBlockIfNeeded();
+    this.serviceTypeAtInit = this.currentExecutingServiceType =
+        abfsOutputStreamContext.getIngressServiceType();
+    this.clientHandler = abfsOutputStreamContext.getClientHandler();
+    createIngressHandler(serviceTypeAtInit,
+        abfsOutputStreamContext.getBlockFactory(), bufferSize);
+    createBlockIfNeeded(position);
   }
 
+  /**
+   * Retrieves the current ingress handler.
+   *
+   * @return the current {@link AzureIngressHandler}.
+   */
+  public AzureIngressHandler getIngressHandler() {
+    return ingressHandler;
+  }
+
+  /**
+   * Creates an ingress handler based on the provided service type and other parameters.
+   *
+   * @param serviceType the type of service, either BLOB or DFS.
+   * @param blockFactory the factory to create data blocks.
+   * @param bufferSize the buffer size for the ingress handler.
+   * @return the created {@link AzureIngressHandler}.
+   * @throws IOException if there is an error creating the ingress handler.
+   */
+  private AzureIngressHandler createIngressHandler(AbfsServiceType serviceType,
```

Comment:
lets have synchronized, as the write to server would happen in parallel, and this method would be called in parallel.

Reply:
Also, there would outputStream.write() which would be needing ingressHandler to create abfsBlock.

Reply:
taken

Reply:
again read the code, we would need some intelligence here as we have a problem here:


there is a synchronized block from the APIs on AbfsOutputStream, and then this sync block from a parallel worker thread , we will have a deadlock here. As if the os.flushInternal is called, it will take a sync block and wait on worker threads to get over, and to take fallback, worker thread would not be able to take sync lock and there will be a deadlock.

Running testCreateOverDfsAppendOverBlob helped understand this issue. Great test!


I would suggest, we have a lock for the worker threads separate from AbfsoutputStream class lock

Reply:
synchronizing is good, but we should also ensure that createIngressHandler is not getting executed by multiple threads.

There should be a check inside createIngressHandler which can say that one of the parallel thread has done the switching, and no more switching can happen.

Reply:
synchronizing is good, but we should also ensure that createIngressHandler is not getting executed by multiple threads.

There should be a check inside createIngressHandler which can say that one of the parallel thread has done the switching, and no more switching can happen.

Reply:
The check in switchHandler :- if (serviceTypeAtInit != currentExecutingServiceType) {
      return;
    } already takes care of this right

Reply:
nice catch, have tried to handle this. Would like you to see if this looks good ?